diff -r 20b23a4880ee config.sub
--- a/config.sub	Thu Jul 01 05:18:32 2010 +0000
+++ b/config.sub	Thu Jul 01 05:32:21 2010 +0000
@@ -1315,7 +1315,7 @@
 	      | -os2* | -vos* | -palmos* | -uclinux* | -nucleus* \
 	      | -morphos* | -superux* | -rtmk* | -rtmk-nova* | -windiss* \
 	      | -powermax* | -dnix* | -nx6 | -nx7 | -sei* | -dragonfly* \
-	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es*)
+	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es* | -microcosm*)
 	# Remember, each alternative MUST END IN *, to match a version number.
 		;;
 	-qnx*)
diff -r 20b23a4880ee gcc/config.gcc
--- a/gcc/config.gcc	Thu Jul 01 05:18:32 2010 +0000
+++ b/gcc/config.gcc	Thu Jul 01 05:32:21 2010 +0000
@@ -604,6 +604,10 @@
     *) echo 'Unknown thread configuration for VxWorks'; exit 1 ;;
   esac
   ;;
+*-*-microcosm)
+  default_use_cxa_atexit=yes
+  use_gcc_stdint=wrap
+  ;;
 *-*-elf)
   # Assume that newlib is being used and so __cxa_atexit is provided.
   default_use_cxa_atexit=yes
@@ -1106,6 +1110,14 @@
 	tm_file="${tm_file} i386/unix.h i386/att.h dbxelf.h elfos.h newlib-stdint.h i386/i386elf.h i386/x86-64.h"
 	tmake_file="${tmake_file} i386/t-i386elf t-svr4"
 	;;
+i[34567]86-*-microcosm)
+        tm_file="${tm_file} i386/unix.h i386/att.h dbxelf.h elfos.h newlib-stdint.h i386/microcosm.h"
+        tmake_file="${tmake_file} i386/t-microcosm t-svr4"
+        ;;
+x86_64-*-microcosm*)
+        tm_file="${tm_file} i386/unix.h i386/att.h dbxelf.h elfos.h newlib-stdint.h i386/x86-64.h i386/microcosm.h"
+        tmake_file="${tmake_file} i386/t-microcosm t-svr4"
+        ;;
 i[34567]86-*-freebsd*)
 	tm_file="${tm_file} i386/unix.h i386/att.h dbxelf.h elfos.h ${fbsd_tm_file} i386/freebsd.h"
 	;;
diff -r 20b23a4880ee gcc/config/i386/microcosm.h
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/gcc/config/i386/microcosm.h	Thu Jul 01 05:32:21 2010 +0000
@@ -0,0 +1,71 @@
+/* Definitions for Microcosm
+   Copyright (C) 2010 Gregor Richards
+
+This file is part of GCC.
+
+GCC is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation; either version 3, or (at your option)
+any later version.
+
+GCC is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with GCC; see the file COPYING3.  If not see
+<http://www.gnu.org/licenses/>.  */
+
+#undef  TARGET_VERSION
+#define TARGET_VERSION fprintf (stderr, " (i386/x86_64 Microcosm)");
+
+#undef  ASM_COMMENT_START
+#define ASM_COMMENT_START "#"
+
+#undef  ASM_APP_ON
+#define ASM_APP_ON "#APP\n"
+
+#undef  ASM_APP_OFF
+#define ASM_APP_OFF "#NO_APP\n"
+
+#undef  DBX_REGISTER_NUMBER
+#define DBX_REGISTER_NUMBER(n) \
+  (TARGET_64BIT ? dbx64_register_map[n] : svr4_dbx_register_map[n])
+
+#define TARGET_OS_CPP_BUILTINS()		\
+  do						\
+    {						\
+	builtin_define ("__MICROCOSM__");	\
+    }						\
+  while (0)
+
+#undef	LINK_SPEC
+#define LINK_SPEC "%{shared:-shared} \
+  %{!shared: \
+      %{!static: \
+	%{rdynamic:-export-dynamic} \
+	%{!dynamic-linker:-dynamic-linker /usr/bin/microcosm-ld}} \
+	%{static:-static}}"
+
+#undef ENDFILE_SPEC
+#define ENDFILE_SPEC "%{shared|pie:crtendS.o%s;:crtend.o%s} crtn.o%s"
+
+#undef STARTFILE_SPEC
+#define STARTFILE_SPEC "%{!shared: \
+			 %{!symbolic: \
+			  %{pg:gcrt0.o%s}%{!pg:%{p:mcrt0.o%s}%{!p:crt0.o%s}}}}\
+                        crti.o%s\
+			%{shared|pie:crtbeginS.o%s;:crtbegin.o%s}"
+
+#if defined(HAVE_LD_EH_FRAME_HDR)
+#define LINK_EH_SPEC "%{!static:--eh-frame-hdr} "
+#endif
+
+#ifdef HAVE_LD_AS_NEEDED
+#define USE_LD_AS_NEEDED 1
+#endif
+
+#undef	LIB_SPEC
+#define LIB_SPEC \
+  "-lc -lmicrocosm -lmicrocosm_posix"
diff -r 20b23a4880ee gcc/config/i386/t-microcosm
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/gcc/config/i386/t-microcosm	Thu Jul 01 05:32:21 2010 +0000
@@ -0,0 +1,4 @@
+# For svr4 we build crtbegin.o and crtend.o which serve to add begin and
+# end labels to the .ctors and .dtors section when we link using gcc.
+
+EXTRA_PARTS=crtbegin.o crtend.o crtbeginS.o crtendS.o
diff -r 20b23a4880ee libgcc/config.host
--- a/libgcc/config.host	Thu Jul 01 05:18:32 2010 +0000
+++ b/libgcc/config.host	Thu Jul 01 05:32:21 2010 +0000
@@ -278,6 +278,10 @@
 	;;
 x86_64-*-elf*)
 	;;
+i[34567]86-*-microcosm*)
+	;;
+x86_64-*-microcosm*)
+	;;
 i[34567]86-*-freebsd*)
 	;;
 x86_64-*-freebsd*)
